File: zeus-web/src/api/ft8-settings.ts
// SPDX-License-Identifier: GPL-2.0-or-later
//
// Zeus — OpenHPSDR Protocol-1 / Protocol-2 client.
// Copyright (C) 2025-2026 Brian Keating (EI6LF),
//                         Douglas J. Cerrato (KB2UKA),
//                         Christian Suarez (N9WAR), and contributors.
//
// Zeus is distributed WITHOUT ANY WARRANTY; see the GNU General Public
// License for details.
//
// REST client for the persisted FT8/FT4 workspace preferences (the curated
// WSJT-X/JTDX KEEP set that is pure behaviour/UI, not radio control). Backed by
// the server (zeus-prefs.db) so the operator's choices survive desktop restarts.
// Defaults mirror Zeus.Contracts.Ft8Settings exactly — surfacing existing engine
// knobs, never changing a default an operator already feels. TX still requires an
// explicit arm; none of these flags transmit on their own.

import { ApiError } from './client';

export type Ft8Settings = {
  // TX & auto-sequence
  autoSequence: boolean;
  callFirst: boolean;
  holdTxFreq: boolean;
  disableTxAfter73: boolean;
  /** 0 = 1st (even), 1 = 2nd (odd). */
  defaultTxSlot: number;
  defaultTxOffsetHz: number;
  // Advanced sequence flags (default off to match WSJT-X)
  rr73InsteadOfRrr: boolean;
  skipGrid: boolean;
  /** Caller max retries before giving up (0 = unlimited). */
  callerMaxRetries: number;
  // Macros
  cqMessage: string;
  cqDxMessage: string;
  freeTextMacro: string;
  // Decode
  /** Decode depth as passes, matching the engine scale (1 = Normal/floor,
   *  >1 = Deep/multi). Default 3 mirrors the engine. 1..4. */
  decodePasses: number;
  showOnlyCq: boolean;
  hideWorkedBefore: boolean;
  // Logging
  autoLog: boolean;
  promptBeforeLog: boolean;
  clearDxAfterLog: boolean;
  reportToComment: boolean;
};

/** Defaults must match Zeus.Contracts.Ft8Settings defaults exactly. */
export const FT8_SETTINGS_DEFAULTS: Ft8Settings = {
  autoSequence: true,
  callFirst: false,
  holdTxFreq: false,
  disableTxAfter73: true,
  defaultTxSlot: 0,
  defaultTxOffsetHz: 1500,
  rr73InsteadOfRrr: true,
  skipGrid: false,
  callerMaxRetries: 0,
  cqMessage: 'CQ',
  cqDxMessage: 'CQ DX',
  freeTextMacro: '',
  decodePasses: 3,
  showOnlyCq: false,
  hideWorkedBefore: false,
  autoLog: true,
  promptBeforeLog: false,
  clearDxAfterLog: true,
  reportToComment: false,
};

function normalize(raw: unknown): Ft8Settings {
  const r = (raw ?? {}) as Record<string, unknown>;
  const d = FT8_SETTINGS_DEFAULTS;
  const bool = (v: unknown, f: boolean) => (typeof v === 'boolean' ? v : f);
  const num = (v: unknown, f: number) => (typeof v === 'number' && Number.isFinite(v) ? v : f);
  const str = (v: unknown, f: string) => (typeof v === 'string' ? v : f);
  return {
    autoSequence: bool(r.autoSequence, d.autoSequence),
    callFirst: bool(r.callFirst, d.callFirst),
    holdTxFreq: bool(r.holdTxFreq, d.holdTxFreq),
    disableTxAfter73: bool(r.disableTxAfter73, d.disableTxAfter73),
    defaultTxSlot: num(r.defaultTxSlot, d.defaultTxSlot) === 0 ? 0 : 1,
    defaultTxOffsetHz: num(r.defaultTxOffsetHz, d.defaultTxOffsetHz),
    rr73InsteadOfRrr: bool(r.rr73InsteadOfRrr, d.rr73InsteadOfRrr),
    skipGrid: bool(r.skipGrid, d.skipGrid),
    callerMaxRetries: num(r.callerMaxRetries, d.callerMaxRetries),
    cqMessage: str(r.cqMessage, d.cqMessage),
    cqDxMessage: str(r.cqDxMessage, d.cqDxMessage),
    freeTextMacro: str(r.freeTextMacro, d.freeTextMacro),
    decodePasses: num(r.decodePasses, d.decodePasses),
    showOnlyCq: bool(r.showOnlyCq, d.showOnlyCq),
    hideWorkedBefore: bool(r.hideWorkedBefore, d.hideWorkedBefore),
    autoLog: bool(r.autoLog, d.autoLog),
    promptBeforeLog: bool(r.promptBeforeLog, d.promptBeforeLog),
    clearDxAfterLog: bool(r.clearDxAfterLog, d.clearDxAfterLog),
    reportToComment: bool(r.reportToComment, d.reportToComment),
  };
}

async function jsonFetch(input: RequestInfo, init: RequestInit | undefined): Promise<Ft8Settings> {
  const res = await fetch(input, init);
  if (!res.ok) throw new ApiError(res.status, `${res.status} ${res.statusText}`);
  return normalize((await res.json()) as unknown);
}

export function getFt8Settings(signal?: AbortSignal): Promise<Ft8Settings> {
  return jsonFetch('/api/ft8/settings', { signal });
}

export function postFt8Settings(settings: Ft8Settings, signal?: AbortSignal): Promise<Ft8Settings> {
  return jsonFetch('/api/ft8/settings', {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(settings),
    signal,
  });
}

File: Zeus.Contracts/Ft8Settings.cs
// SPDX-License-Identifier: GPL-2.0-or-later
//
// Zeus — OpenHPSDR Protocol-1 / Protocol-2 client.
// Copyright (C) 2025-2026 Brian Keating (EI6LF),
//                         Douglas J. Cerrato (KB2UKA),
//                         Christian Suarez (N9WAR), and contributors.
//
// Zeus is distributed WITHOUT ANY WARRANTY; see the GNU General Public
// License for details.
//
// Ft8Settings — the operator's persisted FT8/FT4 workspace preferences (the
// curated WSJT-X/JTDX KEEP set that is purely behaviour/UI, not radio control).
// Persisted server-side (zeus-prefs.db) so they survive desktop restarts — the
// frontend's old localStorage was port-scoped on the webview and lost on every
// launch. Operator IDENTITY (callsign/grid) is NOT here: it lives in the shared
// OperatorIdentity so spotting/FreeDV/TX all read one source.

namespace Zeus.Contracts;

/// <summary>
/// Persisted FT8/FT4 behaviour preferences surfaced on the FT8 Settings page.
/// Defaults are chosen so the first session behaves EXACTLY as it did before this
/// page existed (auto-sequence on, RR73 ack, disable-after-73 on, 3 decode
/// passes) — nothing an operator already feels changes until they touch a
/// control. Most flags are wired to the auto-sequence controller / macros / log
/// path; a couple (SkipGrid, ClearDxAfterLog) are persisted but not yet consumed
/// and are surfaced disabled ("coming soon") on the page. TX still requires an
/// explicit arm; none of these flags transmit on their own.
/// </summary>
public sealed record Ft8Settings(
    // ── TX & auto-sequence ──────────────────────────────────────────────────
    /// <summary>Master auto-sequence on/off. TX still requires ENABLE/arm.</summary>
    bool AutoSequence = true,
    bool CallFirst = false,
    bool HoldTxFreq = false,
    bool DisableTxAfter73 = true,
    /// <summary>Default TX slot: 0 = 1st (even), 1 = 2nd (odd).</summary>
    int DefaultTxSlot = 0,
    /// <summary>Default TX audio offset (Hz). Bounded on write.</summary>
    int DefaultTxOffsetHz = 1500,
    // Advanced sequence flags. RR73 ack is ON by default to match the engine's
    // pre-settings behaviour (the sequencer keys RR73) and modern WSJT-X FT8.
    bool Rr73InsteadOfRrr = true,
    bool SkipGrid = false,
    /// <summary>Caller max retries before giving up (0 = unlimited).</summary>
    int CallerMaxRetries = 0,
    // ── Macros ──────────────────────────────────────────────────────────────
    string CqMessage = "CQ",
    string CqDxMessage = "CQ DX",
    string FreeTextMacro = "",
    // ── Decode ──────────────────────────────────────────────────────────────
    /// <summary>
    /// Decode depth as passes, matching the engine scale (1 = Normal/floor,
    /// &gt;1 = Deep/multi). Default 3 mirrors Ft8Service.DecodePasses so the first
    /// session decodes exactly as deep as before this page existed. 1..4.
    /// </summary>
    int DecodePasses = 3,
    bool ShowOnlyCq = false,
    bool HideWorkedBefore = false,
    // ── Logging ─────────────────────────────────────────────────────────────
    bool AutoLog = true,
    bool PromptBeforeLog = false,
    bool ClearDxAfterLog = true,
    bool ReportToComment = false)
{
    public const int MinOffsetHz = 200;
    public const int MaxTxOffsetHz = 4000;
    public const int MaxMacroLength = 13;     // FT8 free-text is 13 chars
    public const int MinPasses = 1;
    public const int MaxPasses = 4;

    /// <summary>
    /// Clamp/trim so a hand-crafted POST or stale row can't push out-of-range
    /// values: offset bounded, passes 1..4, slot 0/1, macros trimmed/capped,
    /// retries non-negative.
    /// </summary>
    public Ft8Settings Normalized() => this with
    {
        DefaultTxSlot = DefaultTxSlot == 0 ? 0 : 1,
        DefaultTxOffsetHz = Math.Clamp(DefaultTxOffsetHz, MinOffsetHz, MaxTxOffsetHz),
        CallerMaxRetries = Math.Max(0, CallerMaxRetries),
        DecodePasses = Math.Clamp(DecodePasses, MinPasses, MaxPasses),
        CqMessage = Cap(CqMessage, 32),
        CqDxMessage = Cap(CqDxMessage, 32),
        FreeTextMacro = Cap(FreeTextMacro, MaxMacroLength),
    };

    private static string Cap(string? s, int max)
    {
        var t = (s ?? "").Trim();
        return t.Length > max ? t[..max] : t;
    }
}
